File: apps/api/app/api/v1/endpoints/app_config.py
"""
FitNudge V2 - App Configuration Endpoint

Public endpoint to fetch dynamic app configuration values.
No authentication required for public configs.
"""

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel
from typing import Dict, Optional, List

from app.core.database import get_supabase_client
from app.services.logger import logger

router = APIRouter()


class AppConfigItem(BaseModel):
    """Single configuration item"""

    key: str
    value: str
    category: str


class AppConfigResponse(BaseModel):
    """Response containing all public configuration values"""

    config: Dict[str, str]  # Key-value pairs for easy access
    items: List[AppConfigItem]  # Full items with category info


@router.get("/public", response_model=AppConfigResponse)
async def get_public_config():
    """
    Get all public app configuration values.

    No authentication required.

    Returns configuration as both:
    - A flat key-value dict for easy access
    - Full items with category info for filtering
    """
    try:
        supabase = get_supabase_client()

        result = (
            supabase.table("app_config")
            .select("key, value, category")
            .eq("is_public", True)
            .execute()
        )

        items = result.data or []

        # Transform to both formats
        config_dict = {item["key"]: item["value"] for item in items}
        config_items = [
            AppConfigItem(
                key=item["key"], value=item["value"], category=item["category"]
            )
            for item in items
        ]

        return AppConfigResponse(config=config_dict, items=config_items)

    except Exception as e:
        logger.error(f"Failed to fetch app config: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch app configuration",
        )


@router.get("/public/{category}", response_model=AppConfigResponse)
async def get_public_config_by_category(category: str):
    """
    Get public app configuration values filtered by category.

    No authentication required.

    Categories:
    - app_store_urls: iOS and Android store URLs
    - external_urls: Privacy policy, terms, help center, etc.
    """
    try:
        supabase = get_supabase_client()

        result = (
            supabase.table("app_config")
            .select("key, value, category")
            .eq("is_public", True)
            .eq("category", category)
            .execute()
        )

        items = result.data or []

        config_dict = {item["key"]: item["value"] for item in items}
        config_items = [
            AppConfigItem(
                key=item["key"], value=item["value"], category=item["category"]
            )
            for item in items
        ]

        return AppConfigResponse(config=config_dict, items=config_items)

    except Exception as e:
        logger.error(f"Failed to fetch app config for category {category}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch app configuration",
        )

File: apps/mobile/src/services/api/appConfig.ts
/**
 * App Config API Service
 *
 * Fetches dynamic app configuration values from the backend.
 * These values can be updated by admin without requiring an app release.
 *
 * No authentication required for public configs.
 */

import { BaseApiService } from "./base";
import { ROUTES } from "@/lib/routes";

export interface AppConfigItem {
  key: string;
  value: string;
  category: string;
}

export interface AppConfigResponse {
  config: Record<string, string>; // Key-value pairs for easy access
  items: AppConfigItem[]; // Full items with category info
}

/**
 * Typed config keys for type-safe access
 */
export type AppConfigKey =
  // App Store URLs
  | "ios_app_store_url"
  | "android_play_store_url"
  // External URLs
  | "privacy_policy_url"
  | "terms_of_service_url"
  | "help_center_url"
  | "tally_feedback_url"
  | "tally_bug_url"
  | "tawk_chat_url"
  | "contact_email";

export type AppConfigCategory = "app_store_urls" | "external_urls" | "feature_flags";

class AppConfigApiService extends BaseApiService {
  /**
   * Get all public app configuration values
   */
  async getPublicConfig(): Promise<AppConfigResponse> {
    const response = await this.get<AppConfigResponse>(ROUTES.APP_CONFIG.PUBLIC);
    return response.data ?? { config: {}, items: [] };
  }

  /**
   * Get public config values filtered by category
   */
  async getConfigByCategory(category: AppConfigCategory): Promise<AppConfigResponse> {
    const response = await this.get<AppConfigResponse>(ROUTES.APP_CONFIG.BY_CATEGORY(category));
    return response.data ?? { config: {}, items: [] };
  }
}

// Export singleton instance
export const appConfigService = new AppConfigApiService();
